### PR TITLE
feat(compiler): Add extended diagnostic to warn when text attributes …

### LINKED ISSUE #46137: Add extended diagnostic to warn when an attribute is not interpreted as a binding
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

Angular does not properly bind an attribute when using `attr.foo` syntax if there is no interpolation present in the value.

For example, this does not work:

`<div attr.foo="bar"></div> <!-- erroneously keeps the attribute as "attr.foo" -->`

This DOES work:

`<div attr.foo="{{myProp}}"></div>`

This also works:

`<div [attr.foo]="'bar'"></div>`

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-ivy-ybks1w?file=src%2Fapp%2Fapp.component.html,src%2Fapp%2Fapp.module.ts

### Please provide the exception or error you saw

```true
The attribute is not set properly on the element using `attr.foo` syntax when no interpolation is present.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
The reproduction link is using Angular 13, but I also encountered this in my project which produces the following output for `ng version`:


Angular CLI: 12.2.16
Node: 14.19.3
Package Manager: npm 6.14.17
OS: darwin x64

Angular: 12.2.16
... animations, cli, common, compiler, compiler-cli, core
... elements, forms, language-service, localize
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1202.16
@angular-devkit/build-angular   12.2.16
@angular-devkit/core            12.2.16
@angular-devkit/schematics      12.2.16
@angular/cdk                    12.2.13
@schematics/angular             12.2.16
ng-packagr                      12.2.7
rxjs                            6.6.3
typescript                      4.3.5
webpack                         5.50.0
```
```


### Anything else?

I encountered this in my project previously on Angular 11 as well.

### PR DESCRIPTION
**Pull request by atscott**
…are intended to be bindings

https://angular.io/guide/attribute-binding#attribute-class-and-style-bindings

Angular supports `attr.`, `style.`, and `class.` binding prefixes to
bind attributes, styles, and classes. If the key does not have the
binding syntax `[]` or the value does not have an interpolation `{{}}`,
the attribute will not be interpreted as a binding.

This diagnostic warns the user when the attributes listed above will not
be interpreted as bindings.

resolves #46137

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      const expectedKey = `[${name}]`;
+      const expectedValue =
+          node.value === 'true' || node.value === 'false' ? node.value : `'${node.value}'`;
+      errorString = boundSyntax.replace(
+          name,
+          `Attribute, style, and class bindings should be enclosed with square braces: '${
+              expectedKey}="${expectedValue}"'.`);
+    }
+    const diagnostic = ctx.makeTemplateDiagnostic(
+        node.sourceSpan, `${errorString} Find more at https://angular.io/guide/TODO`);
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts:57]: I think you can drop the `Find more at ...`, I believe it should be added automatically if the error code is registered in the `COMPILER_ERRORS_WITH_GUIDES` set.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/text_attribute_not_binding/text_attribute_not_binding_spec.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      expect(diags[0].code).toBe(ngErrorCode(ErrorCode.TEXT_ATTRIBUTE_NOT_BINDING));
+      expect(getSourceCodeForDiagnostic(diags[0])).toBe(`class.blue="true"`);
+    });
+
+    it('should produce an attribute binding warning', () => {
+      const fileName = absoluteFrom('/main.ts');
+      const {program, templateTypeChecker} = setup([{
+        fileName,
+        templates: {
+          'TestCmp': `<div attr.id="bar"></div>`,
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/text_attribute_not_binding/text_attribute_not_binding_spec.ts:54]: Could you plz add a quick test with an attribute without a value, like this:
```
'TestCmp': `<div attr.readonly></div>`,
```

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    let errorString: string;
+    if (name.startsWith('attr.')) {
+      const staticAttr = name.replace('attr.', '');
+      errorString = `Static attributes should be written without the 'attr.' prefix: ${
+          staticAttr}="${node.value}".`;
+
+    } else {
+      const expectedKey = `[${name}]`;
+      const expectedValue =
+          node.value === 'true' || node.value === 'false' ? node.value : `'${node.value}'`;
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts:50]: nit: for better readability:
```suggestion
          (node.value === 'true' || node.value === 'false') ? node.value : `'${node.value}'`;
```

Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts:50]: Btw, could you plz also add a quick comment to describe why do we special-case `true` and `false` (but not numbers, `null`, `undefined`, etc)?

**Review by AndrewKushnir — COMMENTED** [review]
Looks great, just a few minor comments.

**Comment by AndrewKushnir** [comment]
@atscott quick question: do you think we should try to include the check for #36256 in the same diagnostics (as a followup PR) or create a new one?

**Comment by atscott** [comment]
> @atscott quick question: do you think we should try to include the check for #36256 in the same diagnostics (as a followup PR) or create a new one?

Potentially. Would it be limited to just `.px` suffix on `attr.` prefixes? Or warn whenever we encounter a `.` after the `attr.`?

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/text_attribute_not_binding/text_attribute_not_binding_spec.ts (hunk 2 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/text_attribute_not_binding/text_attribute_not_binding_spec.ts:54]: Would you expect this diagnostic to warn in this case? It currently does not.

**Comment by AndrewKushnir** [comment]
> Potentially. Would it be limited to just .px suffix on attr. prefixes? Or warn whenever we encounter a . after the attr.?

We can probably start with `.px` and `.%` suffixes (what is typically used for styles) and expand later if needed.

**Review by AndrewKushnir — APPROVED** [review]
Looks great, thanks for addressing the feedback! 👍

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import {TemplateCheckFactory, TemplateCheckWithVisitor, TemplateContext} from '../../api';
+
+/**
+ * Ensures that attributes that have the "special" angular binding prefix (attr., style., and
+ * class.) are interpreted as bindings. For example, `<div attr.id="my-id"></div>` will not
+ * interperet this as an `AttributeBinding` to `id` but rather just a `TmplAstTextAttribute`. This
+ * is likely not the intent of the developer. Instead, the intent is likely to have the `id` be set
+ * to 'my-id'.
+ */
+class TextAttributeNotBindingSpec extends
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts:23]: Is `Spec` here an abbreviation for `Special`? We should just spell it out in that case - I was confused why there was a test here

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api, fw-compiler

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts (hunk 4 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts:23]: This is unintentional. I think when I was bikeshedding names I was going to use "Special" somewhere but then didn't and this artifact remained.

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: public-api

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts (hunk 4 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts:23]: Or I was reading the `_spec.ts` file name and just had this repeated in my head 😆

**Comment by atscott** [comment]
This PR was merged into the repository by commit 6f11a580406877e440c43df31fae3d5f120cafed.

**Pull request closed without merging**